tap_file/client.py

```diff
@@ -2,14 +2,136 @@
 
 from __future__ import annotations
 
-from typing import Iterable
+import re
+from functools import cached_property
+from pathlib import Path
+from typing import Any, Generator, Iterable
 
+import fsspec
 from singer_sdk.streams import Stream
 
 
 class FileStream(Stream):
     """Stream class for File streams."""
 
+    @cached_property
+    def filesystem(self) -> fsspec.AbstractFileSystem:  # noqa: PLR0911
+        """A fsspec filesytem.
+
+        TODO: Move this logic to an external class if support for further protocols is
+        added.
+
+        Raises:
+            ValueError: If the supplied protocol is not supported.
+
+        Returns:
+            A fileystem object of the appropriate type for the user-supplied protocol.
+        """
+        protocol = self.config["protocol"]
+
+        if protocol == "file":
+            return fsspec.filesystem("file")
+        if protocol == "s3":
+            cache_filepath = self.config.get("cache_filepath", None)
+
+            # A user specified anonymous connection overrides all else, allowing for a
+            # uncredentialed requests even when credentials are available.
+            if self.config["s3_anonymous_connection"]:
+                if cache_filepath:
+                    return fsspec.filesystem(
+                        "filecache",
+                        target_protocol="s3",
+                        target_options={"anon": True},
+                        cache_storage=cache_filepath,
+                    )
+                self.logger.warning(
```

Comment:
Hmm, what is the benefit to having this be configurable? Does it save space? I think at a minimum we should just default to the OS's temp directory and call it a day. This hits on a concept called "convention over configuration" which is that we should recommend a conventional way of doing this, but if there is value we should allow folks to override the configuration.

I almost wonder if we shouldn't allow a config option here but :shrug:

Reply:
My thinking was that for truly massive buckets, there wouldn't be space to cache everything on the local machine and it would be necessary to process each file individually. I'll switch the default to be the temp directory though, since that's probably a better "sensible default" than not using caching.